src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs

```diff
@@ -113,10 +205,77 @@ private string GetFieldsDirective()
         internal override Task WriteMessageAsync(string message, StreamWriter streamWriter, CancellationToken cancellationToken)
         {
             OnWrite(message);
-            return base.WriteMessageAsync(message, streamWriter, cancellationToken);
+
+            if (message.IndexOf(W3CSeparator, 0, System.StringComparison.InvariantCulture) < 0)
+            {
+                return base.WriteMessageAsync(message, streamWriter, cancellationToken);
+            }
+
+            return WriteW3CMessageAsync(message, streamWriter, cancellationToken);
         }
 
         // Extensibility point for tests
         internal virtual void OnWrite(string message) { }
+
+        public void EnqueueMessage(string[] messages)
+        {
+            if (!_messageQueue.IsAddingCompleted)
+            {
+                try
+                {
+                    _messageQueue.Add(string.Join(W3CSeparator, messages));
```

Comment:
Doing string.Join here & string.Split later means we aren't necessarily saving any allocations versus using the StringBuilder - we should try to think of a way to eliminate this and just write each array element directly to the streamWriter, though admittedly that might be complicated with having to enqueue them into the collection of strings.

Reply:
The queue itself will need to change.

Reply:
@wcontayon still need to make this change in order to actually save allocations vs. what's currently in `main`